api/core/models.py

```diff
@@ -69,89 +86,395 @@ class _AbstractBaseAuditableModel(models.Model):
+        message = UPDATED_MESSAGE.format(
+            model_name=model_name,
+            identity=identity,
+            field=change.field,
+            action=_get_action(self, change.field),
+        )
+
+        if m2m_field := m2m_fields.get(change.field):
+
+            def get_identity(pk):
+                obj = m2m_field.related_model.objects.filter(pk=pk).first()
+                # related model MUST implement get_audit_log_identity
+                return obj.get_audit_log_identity() if obj else "None"
+
+            reverse_field_name = m2m_field.m2m_reverse_field_name()
+            old_values = {
+                through[reverse_field_name]: through for through in change.old
+            }
+            new_values = {
+                through[reverse_field_name]: through for through in change.new
+            }
+            for pk in new_values.keys() - old_values.keys():
+                message += f"; added: {get_identity(pk)}"
+            for pk in old_values.keys() - new_values.keys():
+                message += f"; removed: {get_identity(pk)}"
+            for pk in new_values.keys() & old_values.keys():
+                related_identity = get_identity(pk)
+                for field in old_values[pk].keys() | new_values[pk].keys():
+                    if old_values[pk].get(field) != new_values[pk].get(field):
+                        message += f"; {field} changed: {related_identity}"
+
+        return message
+
+    return "; ".join(describe(change) for change in delta.changes) or None
+
+
+def default_get_delete_log_message(
+    self: _AbstractBaseAuditableModel, history_instance
+) -> str | None:
+    return DELETED_MESSAGE.format(
+        model_name=self.get_audit_log_model_name(history_instance),
+        identity=self.get_audit_log_identity(),
+    )
+
+
 def abstract_base_auditable_model_factory(
-    historical_records_excluded_fields: typing.List[str] = None,
+    related_object_type: RelatedObjectType,
+    unaudited_fields: typing.Sequence[str] | None = None,
+    audited_m2m_fields: typing.Sequence[str] | None = None,
+    audit_create=False,
+    audit_update=False,
+    audit_delete=False,
 ) -> typing.Type[_AbstractBaseAuditableModel]:
     class Base(_AbstractBaseAuditableModel):
         history = HistoricalRecords(
             bases=[BaseHistoricalModel],
-            excluded_fields=historical_records_excluded_fields or [],
+            excluded_fields=unaudited_fields or (),
+            m2m_fields=audited_m2m_fields or (),
             get_user=get_history_user,
             inherit=True,
         )
 
         class Meta:
             abstract = True
 
+    Base.related_object_type = related_object_type
+    if audit_create:
+        Base.get_create_log_message = default_get_create_log_message
+    if audit_update:
+        Base.get_update_log_message = default_get_update_log_message
+    if audit_delete:
+        Base.get_delete_log_message = default_get_delete_log_message
```

Comment:
This dance was to avoid changing the default behaviour of `abstract_base_auditable_model_factory` as many things already call this expecting `_AbstractBaseAuditableModel` and its implementations that return `None`, logging nothing by default. 
All the things I've added want to log create/update/delete by default (except project but maybe it should).
I will see if I can do it more elegantly using mixins.

Reply:
I've replaced this with a mixin implementation - it's a bit verbose but hopefully clearer.

Reply:
Yeah, I don't think that's an intentional requirement. I think having project update would also be beneficial.